### PR TITLE
RavenDB-13978 - fix a race condition where we could end up repeating the same Raven-Replication-Version

### PR DESCRIPTION
**Pull request by grisha-kotler**
https://issues.hibernatingrhinos.com/issue/RavenDB-13978

### CODE REVIEW

File: Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs [hunk 1]
```diff
@@ -16,52 +16,59 @@ namespace Raven.Bundles.Replication.Impl
...
+        }
+
+        public long NextId()
+        {
+            var current = valueHolder;
+            var incrementedCurrent = Interlocked.Increment(ref current.Current);
+
+            while (incrementedCurrent > current.MaxValue)
+            {
+                lock (generatorLock)
```
Comment (ayende) [inline Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs:43]: We probably don't need this, given that we now hold all of this in the ValueHolder

File: Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs [hunk 2]
```diff
@@ -16,52 +16,59 @@ namespace Raven.Bundles.Replication.Impl
...
+                {
+                    current = valueHolder;
```
Comment (ayende) [inline Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs:45]: Need to use volatile read here.

File: Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs [hunk 3]
```diff
@@ -16,52 +16,59 @@ namespace Raven.Bundles.Replication.Impl
...
+                    if (current.Current > current.MaxValue)
+                    {
                         using (var locker = Database.DocumentLock.TryLock(250))
                         {
                             if (locker == null)
                                 continue;
-                            currentMax = new Hodler(GetNextMax());
+
+                            current = valueHolder = GetNextMax(current.MaxValue);
```
Comment (ayende) [inline Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs:53]: valueHolder *must* only be set using `Interlocked` operations

**Review by ayende — CHANGES_REQUESTED** [review]

File: Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs (hunk 3 above)
Comment (grisha-kotler) [reply Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs:53]: This is running under the DocumentLock

File: Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs (hunk 2 above)
Comment (grisha-kotler) [reply Raven.Database/Bundles/Replication/Impl/ReplicationHiLo.cs:45]: valueHolder is volatile

**Review by ppekrol — APPROVED** [review]

**Review by ayende — APPROVED** [review]

**Pull request merged**